src/ALabel.cpp

```diff
@@ -14,30 +15,129 @@ waybar::ALabel::ALabel(const Json::Value& config, const std::string format)
   }
   if (config_["format-alt"].isString()) {
     event_box_.add_events(Gdk::BUTTON_PRESS_MASK);
-    event_box_.signal_button_press_event()
-      .connect(sigc::mem_fun(*this, &ALabel::handleToggle));
+    event_box_.signal_button_press_event().connect(
+        sigc::mem_fun(*this, &ALabel::handleToggle));
+  }
+
+  // configure events' user commands
+  if (config_["on-click"].isString()) {
+    std::string cmd = config_["on-click"].asString();
+    setEventCmd(waybar::EventType::BUTTON_PRESS, cmd);
+  }
+  if (config_["on-scroll-up"].isString()) {
+    std::string cmd = config_["on-scroll-up"].asString();
+    setEventCmd(waybar::EventType::SCROLL_UP, cmd);
+  }
+  if (config_["on-scroll-down"].isString()) {
+    std::string cmd = config_["on-scroll-down"].asString();
+    setEventCmd(waybar::EventType::SCROLL_DOWN, cmd);
   }
 }
 
-auto waybar::ALabel::update() -> void
-{
+auto waybar::ALabel::update() -> void {
   // Nothing here
 }
 
-bool waybar::ALabel::handleToggle(GdkEventButton* const& /*ev*/)
-{
-  alt = !alt;
-  if (alt) {
-    format_ = config_["format-alt"].asString();
-  } else {
-    format_ = default_format_;
+bool waybar::ALabel::handleToggle(GdkEventButton* const& e) {
+  bool ret = false;
+  if (e->button == 1) {
+    ret = forkExec(button_press_cmd_);
+  } else if (e->button == 2) {
+    alt = !alt;
+    if (alt) {
+      format_ = config_["format-alt"].asString();
+    } else {
+      format_ = default_format_;
+    }
+    ret = true;
   }
+
   dp.emit();
+  return ret;
+}
+
+bool waybar::ALabel::forkExec(std::string cmd) {
+  int32_t pid = fork();
+
+  if (pid < 0) {
+    printf("Unable to exec cmd %s, error %s", cmd.c_str(), strerror(errno));
+    return false;
+  }
+
+  // Child executes the command
+  if (!pid) waybar::util::command::exec(cmd);
+
   return true;
 }
 
-std::string waybar::ALabel::getIcon(uint16_t percentage, const std::string& alt)
-{
+bool waybar::ALabel::handleScroll(GdkEventScroll* e) {
+  bool ret = false;
+
+  // Avoid concurrent scroll event
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+    bool direction_up = false;
+
+    if (e->direction == GDK_SCROLL_UP) {
+      direction_up = true;
+    }
+    if (e->direction == GDK_SCROLL_DOWN) {
+      direction_up = false;
+    }
+    if (e->direction == GDK_SCROLL_SMOOTH) {
+      gdouble delta_x, delta_y;
+      gdk_event_get_scroll_deltas(reinterpret_cast<const GdkEvent*>(e),
+                                  &delta_x, &delta_y);
+      if (delta_y < 0) {
+        direction_up = true;
+      } else if (delta_y > 0) {
+        direction_up = false;
+      }
+    }
+
+    if (direction_up)
```

Comment:
Check if user provided both up and scroll cmds :)

Reply:
Sorry, didn't understand. At one time we would be processing only one direction, right?

Reply:
If the user provided only scroll up the scroll down will fail :?

Reply:
in that case, scroll down would be called but since the cmd for it would be "", it would act as a no-op.
Probably we could add a check in forkexec for null commands.

Reply:
Ok my bad ;)